### DISCUSSION TITLE
How to enable anonymous signin in local development?

### DISCUSSION THREAD

**Discussion by ngamsomset** [supabase/supabase#22359] [Questions] [answered]
according to the latest announcement that Supabase included anonymous signin here. https://supabase.com/docs/guides/auth/auth-anonymous#access-control

I wonder how to enable this locally?

This is possible to do with hosted project by go to `setting` in the Dashboard. But local studio doesn't have that setting page.

![CleanShot 2024-04-01 at 14 27 06@2x](https://github.com/supabase/supabase/assets/60306190/6c4e65cd-ca6b-4dbe-b2c2-e34966b96977)

**Comment by endyjasmi** [comment]
I am stucked on the same issues

**Reply by ngamsomset** [reply]
https://github.com/supabase/auth/commit/130df165270c69c8e28aaa1b9421342f997c1ff3#diff-ab56fdad4e2b68148c279d0ee7551a35286f8f32dea65c76dec2fee352c46792R47

see if this helps.

**Reply by endyjasmi** [reply]
After going through the commit.

I have tried adding the following to the `supabase/config.toml`.
```
[auth.external.anonymous_users]
enabled = true
client_id = "anonymous_users_client_id"
secret = "anonymous_users_secret"
```

Verified that the env variable is properly set at docker using `docker inspect {container_id}`.
```
"GOTRUE_EXTERNAL_ANONYMOUS_USERS_ENABLED=true",
"GOTRUE_EXTERNAL_ANONYMOUS_USERS_CLIENT_ID=anonymous_users_client_id",
"GOTRUE_EXTERNAL_ANONYMOUS_USERS_SECRET=anonymous_users_secret",
"GOTRUE_EXTERNAL_ANONYMOUS_USERS_REDIRECT_URI=http://127.0.0.1:54321/auth/v1/callback",
```
Also tried verifying by hitting the REST endpoint;
```
GET /auth/v1/settings HTTP/1.1
Accept: application/json
Host: 127.0.0.1:54321
```

Following is the response
```
{
  "external": {
    "apple": false,
    "azure": false,
    "bitbucket": false,
    "discord": false,
    "facebook": false,
    "figma": false,
    "fly": false,
    "github": false,
    "gitlab": false,
    "google": false,
    "keycloak": false,
    "kakao": false,
    "linkedin": false,
    "linkedin_oidc": false,
    "notion": false,
    "spotify": false,
    "slack": false,
    "workos": false,
    "twitch": false,
    "twitter": false,
    "email": true,
    "phone": true,
    "zoom": false
  },
  "disable_signup": false,
  "mailer_autoconfirm": true,
  "phone_autoconfirm": true,
  "sms_provider": "",
  "mfa_enabled": false,
  "saml_enabled": false
}
```

_Doesn't seem to work._

For now at least, I am implementing the anonymous feature by signing up a new account with random email and random password. If there is a need to persist the account, I will just update the email and password later. 

```
### Sign up anonymously
POST /auth/v1/signup HTTP/1.1
Accept: application/json
Host: 127.0.0.1:54321
Content-Type: application/json

{
  "email": "[random-hash]@anonymous.user",
  "password": "[random-hash]"
}
```

I can verify if the account is anonymous by checking the logged in email with custom domain. The catch will be there are no rate limit on that email authentication.

Hope's this help until we can find a way to enable anonymous login on local development.

**Reply by ngamsomset** [reply]
I was thinking something similar if I have like a fake user for those needed to see as anonymous. so that user doesn't need to be created every time which might flood the database.

**Reply by endyjasmi** [reply]
sound's like a good solution, in my case, I'll need to generate PII (Personally Identifiable Information) for everybody that access the service to generate a persona, hence the anonymous user authentication.

hopefully this get answered someday ~ 🤞🏻

**Reply by devpratico** [reply]
Same question here, need to test locally before going to prod!

**Comment by ngamsomset** [comment]
> sound's like a good solution, in my case, I'll need to generate PII (Personally Identifiable Information) for everybody that access the service to generate a persona, hence the anonymous user authentication.
> 
> hopefully this get answered someday ~ 🤞🏻

I have try the method that I mentioned. It seem to work. But the side effect is not good. If non-user can access one page in the application by using fake user to log in, while they don't have a proper navigation to go to other pages on the app but if they trying to access from the url, they will be able to because fake user still generate a proper token like real user(indeed it's a real user).

**Comment by ngamsomset** [comment]
> After going through the commit.
> 
> I have tried adding the following to the `supabase/config.toml`.
> 
> ```
> [auth.external.anonymous_users]
> enabled = true
> client_id = "anonymous_users_client_id"
> secret = "anonymous_users_secret"
> ```
> 
> Verified that the env variable is properly set at docker using `docker inspect {container_id}`.
> 
> ```
> "GOTRUE_EXTERNAL_ANONYMOUS_USERS_ENABLED=true",
> "GOTRUE_EXTERNAL_ANONYMOUS_USERS_CLIENT_ID=anonymous_users_client_id",
> "GOTRUE_EXTERNAL_ANONYMOUS_USERS_SECRET=anonymous_users_secret",
> "GOTRUE_EXTERNAL_ANONYMOUS_USERS_REDIRECT_URI=http://127.0.0.1:54321/auth/v1/callback",
> ```
> 
> Also tried verifying by hitting the REST endpoint;
> 
> ```
> GET /auth/v1/settings HTTP/1.1
> Accept: application/json
> Host: 127.0.0.1:54321
> ```
> 
> Following is the response
> 
> ```
> {
>   "external": {
>     "apple": false,
>     "azure": false,
>     "bitbucket": false,
>     "discord": false,
>     "facebook": false,
>     "figma": false,
>     "fly": false,
>     "github": false,
>     "gitlab": false,
>     "google": false,
>     "keycloak": false,
>     "kakao": false,
>     "linkedin": false,
>     "linkedin_oidc": false,
>     "notion": false,
>     "spotify": false,
>     "slack": false,
>     "workos": false,
>     "twitch": false,
>     "twitter": false,
>     "email": true,
>     "phone": true,
>     "zoom": false
>   },
>   "disable_signup": false,
>   "mailer_autoconfirm": true,
>   "phone_autoconfirm": true,
>   "sms_provider": "",
>   "mfa_enabled": false,
>   "saml_enabled": false
> }
> ```
> 
> _Doesn't seem to work._
> 
> For now at least, I am implementing the anonymous feature by signing up a new account with random email and random password. If there is a need to persist the account, I will just update the email and password later.
> 
> ```
> ### Sign up anonymously
> POST /auth/v1/signup HTTP/1.1
> Accept: application/json
> Host: 127.0.0.1:54321
> Content-Type: application/json
> 
> {
>   "email": "[random-hash]@anonymous.user",
>   "password": "[random-hash]"
> }
> ```
> 
> I can verify if the account is anonymous by checking the logged in email with custom domain. The catch will be there are no rate limit on that email authentication.
> 
> Hope's this help until we can find a way to enable anonymous login on local development.

I think that configuration is not available yet. According to this.
https://supabase.com/docs/guides/cli/config#auth.external.provider.enabled

**Reply by endyjasmi** [reply]
Yup, but that's what I infer based on the link you gave.

**Comment by yosukesannohe** [comment, marked as answer]
@ngamsomset 

https://supabase.com/docs/guides/cli/config#auth.enable_anonymous_sign_ins

```
# Allow/disallow anonymous sign-ins to your project.
enable_anonymous_sign_ins = true
```

I did it!

**Reply by ngamsomset** [reply]
Woohoo, yessss!
